File: src/main/kotlin/io/github/ephemient/aoc2019/Day11.kt
package io.github.ephemient.aoc2019

import kotlinx.coroutines.runBlocking

class Day11(private val lines: List<String>) {
    private val ints: List<Long> =
        lines.first().splitToSequence(",").map { it.toLong() }.toList()

    @Suppress("LoopWithTooManyJumpStatements")
    private suspend fun walk(start: Boolean): Map<Pair<Int, Int>, Boolean> {
        val grid = mutableMapOf(0 to 0 to start)
        var x = 0
        var y = 0
        var dx = 0
        var dy = 1
        val vm = Intcode(ints.toMutableList())
        val input = suspend { if (grid[x to y] == true) 1L else 0L }
        while (true) {
            grid[x to y] = (vm.getOutput(input) ?: break) != 0L
            if ((vm.getOutput(input) ?: break) != 0L) {
                dy = -dx.also { dx = dy }
            } else {
                dx = -dy.also { dy = dx }
            }
            x += dx
            y += dy
        }
        return grid
    }

    fun part1(): Int = runBlocking { walk(false) }.count()

    fun part2(): String {
        val result = runBlocking { walk(true) }
            .mapNotNullTo(mutableSetOf<Pair<Int, Int>>()) { (key, value) -> key.takeIf { value } }
        val (xs, ys) = result.unzip()
        return (checkNotNull(ys.min())..checkNotNull(ys.max())).reversed().joinToString("\n") { y ->
            (checkNotNull(xs.min())..checkNotNull(xs.max())).joinToString("") { x ->
                if (x to y in result) "\u2593" else "\u2591"
            }
        }
    }
}

File: src/aoc2019/day11.py
from aoc2019.intcode import Intcode
import asyncio
import fileinput


async def walk(mem, start):
    grid = {(0, 0): start}
    x, y = 0, 0
    dx, dy = 0, 1

    async def input():
        return int(grid.get((x, y), False))

    aiter = Intcode(mem, input).__aiter__()
    try:
        while True:
            grid[x, y] = bool(await aiter.__anext__())
            if await aiter.__anext__():
                dx, dy = dy, -dx
            else:
                dx, dy = -dy, dx
            x += dx
            y += dy
    except StopAsyncIteration as _:
        pass
    return grid


def part1(lines):
    '''
    >>> part1(["3,0,5,0,-1,104,1,104,0,3,0,5,0,-1,104,0,104,0,3,0,104,1,104,0,3,0,104,1,104,0,3,0,6,0,-1,104,0,104,1,3,0,104,1,104,0,3,0,104,1,104,0,99"])
    6
    '''
    return len(
        asyncio.run(walk([int(s.strip()) for s in lines[0].split(',')],
                         False)))


def part2(lines):
    result = [
        k for k, v in asyncio.run(
            walk([int(s.strip()) for s in lines[0].split(',')], True)).items()
        if v
    ]
    return '\n'.join(''.join('\u2591\u2593'[int((x, y) in result)]
                             for x in range(min(k[0] for k in result),
                                            max(k[0] for k in result) + 1))
                     for y in range(max(k[1] for k in result),
                                    min(k[1] for k in result) - 1, -1))


parts = (part1, part2)

if __name__ == '__main__':
    lines = list(fileinput.input())
    print(part1(lines))
    print(part2(lines))
